### DISCUSSION TITLE
PostgreSQL Partitioned Tables

### DISCUSSION THREAD

**Discussion by onx2** [diesel-rs/diesel#3819] [Q&A] [answered]
I would like to know if Diesel supports PostgreSQL [partitioned tables](https://www.postgresql.org/docs/current/ddl-partitioning.html). I've seen a few related issues, but there doesn't seem to be a definitive answer (#3201, #2817, #1661).

The example given in the PG docs that I tested with:
```sql
CREATE TABLE measurement (
    city_id         int not null,
    logdate         date not null,
    peaktemp        int,
    unitsales       int
) PARTITION BY RANGE (logdate);
```

Running `diesel setup` appears to work at first, but there isn't actually anything generated in `schema.rs`. Running `diesel migration run` does however present an error message for me:
```bash
Diesel only supports tables with primary keys. Table measurement has no primary key
```
If this is supported, planned, or simply will never be supported, it would be great to get an official answer here as well as a simple example if it is supported.

Thank you in advance for any information that can be provided on this subject and thanks for making such a wonderful ORM 😄

**Comment by weiznich** [comment, marked as answer]
It's really not possible to provide an "official" answer here, as such an answer would depend a lot on what exactly you want to know. You question in asked in a quite broad way.

Let me go through that step by step:

> would like to know if Diesel supports PostgreSQL [partitioned tables](https://www.postgresql.org/docs/current/ddl-partitioning.html).

By default `diesel` (as the crate, not the CLI tool) supports almost all SQL constructs in some way. Sometimes the support is just not exposed by a premade DSL, which means you need to provide your own extension to make something work. In this case I do not see any reason why working with partitioned tables shouldn't be supported given a `table!` definition.

`diesel` as in the CLI tool does only support generating `table!` definitions for tables that contain a primary key in their `information_schema` entries in postgresql. That seems not to be the case for partitioned tables for whatever reason. It's up to discussion if this is a missing feature in diesel-cli, a bug in postgres or something differently.

> If this is supported, planned, or simply will never be supported, it would be great to get an official answer here as well as a simple example if it is supported.

We are always open to accept PR's to add new features, so feel free to provide patches for correctly generating `table!` definitions for partitioned tables. Alternatively it's always possible to discuss funding work of one of the maintainers on a certain feature. Otherwise as this is not a feature that seems to be used or needed by any of the maintainers its not planned to spend time working on it.

**Reply by onx2** [reply]
Thanks for the quick reply. It sounds like it is possible with some customization. This isn't an immediately needed feature but it's nice to know it may be technically possible to use.

> By default diesel (as the crate, not the CLI tool) supports almost all SQL constructs in some way. Sometimes the support is just not exposed by a premade DSL, which means you need to provide your own extension to make something work. In this case I do not see any reason why working with partitioned tables shouldn't be supported given a table! definition.

Would you suggest [ignoring](https://diesel.rs/guides/configuring-diesel-cli#the-filter-field) the partitioned table and providing a [patch](https://diesel.rs/guides/configuring-diesel-cli#the-patch_file-field) as a starting place?

When / if the time comes, I'd very much consider funding or contributing. Thanks for your time!

**Reply by weiznich** [reply]
> Would you suggest [ignoring](https://diesel.rs/guides/configuring-diesel-cli#the-filter-field) the partitioned table and providing a [patch](https://diesel.rs/guides/configuring-diesel-cli#the-patch_file-field) as a starting place?

That would be a solution, although I likely would just put the manual written `table!` definition into a different file.
